### ISSUE TITLE
Hoist ref declarations in pipe expressions

### ISSUE THREAD

**Issue by STRd6** [DanielXMoore/Civet#406]
> *Originally posted in Discord:*

We can create a `type: "HoistableDec"` AST node that is a statement ending with a semi-colon `let ref;` at the front of any expression that needs to use refs. Then we wrap the expression in parens `(ref = f(), {x: ref.x, y: ref.y})`. If the expression is in statement position it ends up like `let ref;(ref = f(), {x: ref.x, y: ref.y})`. If it is in expression position the `HoistableDec` will be pulled up to the nearest containing block statement level (as one of the last process steps, after any step that could create them).

This same mechanism should work for pipe, globs, compound assignments, etc. (#50).

**Comment by edemaine** [comment]
I believe this was done a while ago, though several parts of #50 still need to use this.

**Comment by edemaine** [comment]
Looking at this further, I think this issue is in fact still open.  An example of a bad compilation is

```js
result = (x() ||> y |> z)
↓↓↓
result = (let ref;y((ref = x())),z(ref))
```

**Comment by STRd6** [comment]
Fixed in #651
